helix-view/src/document.rs

```diff
@@ -1692,15 +1711,97 @@ impl Document {
+        let text = self.text();
+
+        // TODO: convert inside server
+        let start =
+            if let Some(start) = lsp_pos_to_pos(text, diagnostic.range.start, offset_encoding) {
+                start
+            } else {
+                log::warn!("lsp position out of bounds - {:?}", diagnostic);
+                return None;
+            };
+
+        let end = if let Some(end) = lsp_pos_to_pos(text, diagnostic.range.end, offset_encoding) {
+            end
+        } else {
+            log::warn!("lsp position out of bounds - {:?}", diagnostic);
+            return None;
+        };
+
+        let severity = diagnostic.severity.map(|severity| match severity {
+            lsp::DiagnosticSeverity::ERROR => Error,
+            lsp::DiagnosticSeverity::WARNING => Warning,
+            lsp::DiagnosticSeverity::INFORMATION => Info,
+            lsp::DiagnosticSeverity::HINT => Hint,
+            severity => unreachable!("unrecognized diagnostic severity: {:?}", severity),
+        });
+
+        if let Some(lang_conf) = lang_conf {
+            if let Some(severity) = severity {
+                if severity < lang_conf.diagnostic_severity {
+                    return None;
+                }
+            }
+        };
+        use helix_core::diagnostic::{DiagnosticTag, NumberOrString};
+
+        let code = match diagnostic.code.clone() {
+            Some(x) => match x {
+                lsp::NumberOrString::Number(x) => Some(NumberOrString::Number(x)),
+                lsp::NumberOrString::String(x) => Some(NumberOrString::String(x)),
+            },
+            None => None,
+        };
+
+        let tags = if let Some(tags) = &diagnostic.tags {
+            let new_tags = tags
+                .iter()
+                .filter_map(|tag| match *tag {
+                    lsp::DiagnosticTag::DEPRECATED => Some(DiagnosticTag::Deprecated),
+                    lsp::DiagnosticTag::UNNECESSARY => Some(DiagnosticTag::Unnecessary),
+                    _ => None,
+                })
+                .collect();
+
+            new_tags
+        } else {
+            Vec::new()
+        };
+
+        Some(Diagnostic {
+            range: Range { start, end },
+            line: diagnostic.range.start.line as usize,
+            message: diagnostic.message.clone(),
+            severity,
+            code,
+            tags,
+            source: diagnostic.source.clone(),
+            data: diagnostic.data.clone(),
+            language_server_id,
+        })
+    }
+
     #[inline]
     pub fn diagnostics(&self) -> &[Diagnostic] {
         &self.diagnostics
     }
 
     pub fn shown_diagnostics(&self) -> impl Iterator<Item = &Diagnostic> + DoubleEndedIterator {
         self.diagnostics.iter().filter(|d| {
-            self.language_servers_with_feature(LanguageServerFeature::Diagnostics)
-                .any(|ls| ls.id() == d.language_server_id)
+            self.language_servers.len() == 0
```

Comment:
This is slightly hacky, there's likely a cleaner and more correct way to do this, but this is probably much simpler (I'm very open to better solutions though).

The issue is, that language servers are initialized and assigned somewhere else (in `helix_view::Editor::launch_language_servers`) after the Document is opened and only, when it is really opened. In the pickers, this isn't the case, but the language server is likely loaded...

Reply:
hmm I am not sure how correct this is, if a language server doesn't have `LanguageServerFeature::Diagnostic` then we shounldn't show the diagnosts in the preview either I think? 

A less hacky (and generally faster) solution could be to remove the `shown_diagnostics` function and instead simply only copy the diagnostics from the editor for the language servers that have the appropiate feature.

That would make code less error prone in general (remiving the common pitfall of using `shown_diagnostics instead of normal diagnostics).

This would necessitate repopulating all document diagnostics on config reload but I think I would be ok with that. Right now it would be a bit messy (you would have to add it to refresh_config in application.rs) but with the event system we could use that meachanisim to move that behaviour to helix_view where it belongs

Reply:
Yeah this is ugly and incorrect, I'm not happy about this either.

And I agree, but the issue is (why `shown_diagnostics` exists at all), that the command `code_action` relies on non-visible diagnostics (I had a case when I implemented that, not sure which language server it was, where filtered diagnostics would hide some code-actions because of this). This can probably be solved by using the editor diagnostics in `code_action` directly though (maybe via the `doc_diagnostics` as described above).

Reply:
you will not be able to use the `doc_diagnostecs` function described above because diagnostics are mapped rough changes in the document while those on `Editor` are not. I think it could be ok if we just don't show codeactions that pertain only to diagnostics if diagnostics are disabled for a LS (I think that wouldn't be surprising UI wise that ifyoou disable diagnistoces that the associated codeactions also go away).

Reply:
Right... thanks for the hint. I guess then it's indeed better to just disable code-actions that are dependent on the diagnostics and remove the `shown_diagnostics`, I've done that in the last commit.